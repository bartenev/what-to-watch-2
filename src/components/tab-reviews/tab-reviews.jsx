import React, {useEffect} from "react";
import PropTypes from "prop-types";
import {MONTHS} from "../../const";
import {connect} from "react-redux";
import {ActionCreator, Operations} from "../../reducer/data/data";
import {getComments} from "../../reducer/data/selectors";

const getTwoDigitNumber = (number) => {
  return number < 10 ? `0${number}` : number;
};

const getReviews = (comments) => {
  return (
    comments.map((comment) => {
      const {user, date, text, rating, id} = comment;
      const {name} = user;
      const year = date.getFullYear();
      const month = date.getMonth();
      const day = date.getDate();

      const dateTime = `${year}-${getTwoDigitNumber(month + 1)}-${getTwoDigitNumber(day)}`;

      return (
        <div key={`${name}-${id}`} className="review">
          <blockquote className="review__quote">
            <p className="review__text">{text}</p>

            <footer className="review__details">
              <cite className="review__author">{name}</cite>
              <time className="review__date" dateTime={dateTime}>{`${MONTHS[month]} ${day}, ${year}`}</time>
            </footer>
          </blockquote>

          <div className="review__rating">{rating}</div>
        </div>
      );
    }));
};

const TabReviews = (props) => {
  const {film, loadComments, comments, resetComments} = props;

  useEffect(() => {
    loadComments(film.id);

    return () => {
      resetComments();
    };
  }, []);

  const firstCol = Math.ceil(comments.length / 2);

  return (
    <div className="movie-card__reviews movie-card__row">
      <div className="movie-card__reviews-col">
        {getReviews(comments.slice(0, firstCol))}
      </div>

      <div className="movie-card__reviews-col">
        {getReviews(comments.slice(firstCol))}
      </div>
    </div>
  );
};

TabReviews.propTypes = {
  film: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    runTime: PropTypes.number.isRequired,
    genre: PropTypes.string.isRequired,
    released: PropTypes.number.isRequired,
    description: PropTypes.string.isRequired,
    director: PropTypes.string.isRequired,
    starring: PropTypes.arrayOf(PropTypes.string.isRequired).isRequired,
    src: PropTypes.shape({
      poster: PropTypes.string.isRequired,
      backgroundImage: PropTypes.string.isRequired,
      previewImage: PropTypes.string.isRequired,
      previewVideo: PropTypes.string.isRequired,
      video: PropTypes.string.isRequired,
    }).isRequired,
    rating: PropTypes.shape({
      score: PropTypes.number.isRequired,
      count: PropTypes.number.isRequired,
    }).isRequired,
    isFavorite: PropTypes.bool.isRequired,
    backgroundColor: PropTypes.string.isRequired,
  }).isRequired,
  loadComments: PropTypes.func.isRequired,
  comments: PropTypes.arrayOf(PropTypes.shape({
    user: PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
    }),
    date: PropTypes.instanceOf(Date),
    text: PropTypes.string.isRequired,
    rating: PropTypes.number.isRequired,
    id: PropTypes.number.isRequired,
  })),
  resetComments: PropTypes.func.isRequired,
};

const mapStateToProps = (state) => ({
  comments: getComments(state),
});

const mapDispatchToProps = (dispatch) => ({
  loadComments(id) {
    dispatch(Operations.loadComments(id));
  },
  resetComments() {
    dispatch(ActionCreator.resetComments());
  }
});

export {TabReviews};

export default connect(mapStateToProps, mapDispatchToProps)(TabReviews);
