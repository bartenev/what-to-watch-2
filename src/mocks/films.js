import {Genres} from "../const";

const getRandomNumber = (max, min = 0) => {
  return min + Math.round(Math.random() * (max - min));
};

const names = [`Vasya`, `Kolya`, `Ivan`, `Petr`, `Genka`, `Cheba`, `Ilya`, `Karina`];

const directors = [
  `Christopher Nolan`,
  `Steven Spielberg`,
  `Quentin Tarantino`,
  `Martin Scorsese`,
  `David Fincher`,
  `Ridley Scott`,
  `Stanley Kubrick`,
  `Robert Zemeckis`,
  `Francis Ford Coppola`,
  `Clint Eastwood`
];

const genres = Object.values(Genres);

const text = `
Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Praesent in nisi condimentum, aliquet nisl at, fermentum neque.
Vivamus ex urna, tincidunt eget feugiat a, luctus ut arcu.
Vivamus elementum nisl eu elit cursus, id luctus justo posuere.
Integer commodo enim eget ullamcorper faucibus. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu.
Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem.
Quisque fringilla lacinia pulvinar. Donec vel semper nunc, ac posuere lorem. Nulla sed tortor nunc.`;

const textItems = text.split(`.\n`);

const getRandomText = () => {
  let randomText = ``;
  for (let i = 0; i < getRandomNumber(5, 1); i++) {
    randomText += textItems[getRandomNumber(textItems.length - 1)];
    randomText += `. `;
  }
  return randomText;
};

const getRandomDate = () => {
  const targetDate = new Date();
  const rangeDays = 100;
  const diffValue = -1 * getRandomNumber(rangeDays, 0);
  targetDate.setDate(targetDate.getDate() + diffValue);
  targetDate.setHours(getRandomNumber(23, 0));
  targetDate.setMinutes(getRandomNumber(59, 0));
  return targetDate;
};

const getReview = () => {
  return {
    name: names[getRandomNumber(names.length - 1)],
    date: getRandomDate(),
    text: getRandomText(),
    rating: getRandomNumber(10),
  };
};

const getRandomElementOfArray = (array) => {
  return array[getRandomNumber(array.length - 1)];
};

const getRandomArrayOfSomething = (getElement, max, min, array = []) => {
  const count = getRandomNumber(max, min);
  let randomArray = [];
  for (let i = 0; i < count; i++) {
    randomArray.push(getElement(array));
  }

  return randomArray;
};

const filmsNames = [`Bohemian Rhapsody`, `Macbeth`, `Aviator`, `We need to talk about Kevin`, `What We Do in the Shadows`,
  `Revenant`, `Johnny English`, `Shutter Island`, `Pulp Fiction`, `No Country for Old Men`, `Snatch`, `Moonrise Kingdom`,
  `Seven Years in Tibet`, `Midnight Special`, `War of the Worlds`, `Dardjeeling Limited`, `Orlando`, `Mindhunter`, `Midnight Special`];

const getFilms = () => {
  return filmsNames.map((filmName) => {
    const score = getRandomNumber(10, 0);
    let level = `Excellent`;

    if (score <= 6) {
      level = `Bad`;
    } else if (score <= 8) {
      level = `Good`;
    }

    return ({
      title: filmName,
      src: {
        poster: `img/${filmName.toLowerCase().replace(/ /g, `-`)}.jpg`,
        preview: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        video: ``,
      },
      director: getRandomElementOfArray(directors),
      starring: getRandomArrayOfSomething(getRandomElementOfArray, 6, 1, directors),
      reviews: getRandomArrayOfSomething(getReview, 6, 0),
      runTime: getRandomNumber(300, 1),
      genre: getRandomElementOfArray(genres),
      released: getRandomNumber(2021, 1950),
      description: getRandomText(),
      rating: {
        score,
        level,
        count: getRandomNumber(10000),
      },
    });
  });
};

const films = getFilms();

export default films;
