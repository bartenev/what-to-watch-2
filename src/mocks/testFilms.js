export const testFilms =
  [
    {
      id: 3,
      title: `Pulp Fiction`,
      src: {
        poster: `img/pulp-fiction.jpg`,
        backgroundImage: `img/pulp-fiction.jpg`,
        previewImage: `img/pulp-fiction.jpg`,
        previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
      },
      director: `Christopher Nolan`,
      starring: [
        `Quentin Tarantino`,
        `David Fincher`,
      ],
      runTime: 240,
      genre: `Comedies`,
      released: 1956,
      description: `Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu. Integer commodo enim eget ullamcorper faucibus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. Integer commodo enim eget ullamcorper faucibus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. `,
      rating: {
        score: 5,
        count: 783,
      },
      isFavorite: true,
      backgroundColor: `#123456`,
    },
    {
      id: 7,
      title: `Macbeth`,
      src: {
        poster: `img/macbeth.jpg`,
        backgroundImage: `img/macbeth.jpg`,
        previewImage: `img/macbeth.jpg`,
        previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
        video: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`,
      },
      director: `Christopher Nolan`,
      starring: [
        `Robert Zemeckis`,
        `Christopher Nolan`,
        `Clint Eastwood`,
      ],
      runTime: 220,
      genre: `Crime`,
      released: 1966,
      description: `Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. Vivamus ex urna, tincidunt eget feugiat a, luctus ut arcu. Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu. `,
      rating: {
        score: 4,
        count: 6354,
      },
      isFavorite: false,
      backgroundColor: `#987654`,
    },
  ];
