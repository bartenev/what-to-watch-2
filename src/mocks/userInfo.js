import {AuthorizationStatus} from "../reducer/user/user";

export const UserInfo = {
  authorizationStatus: AuthorizationStatus.AUTH,
  userInfo: {
    id: 5,
    email: `nikita@gmail.com`,
    name: `nikita`,
    avatar: `url`,
  }
};
