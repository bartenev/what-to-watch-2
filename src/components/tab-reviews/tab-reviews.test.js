import React from "react";
import renderer from "react-test-renderer";
import {TabReviews} from "./tab-reviews";
import {testFilms} from "../../mocks/testFilms";

const comments = [
  {
    user: {
      id: 6,
      name: `Nikita`,
    },
    date: new Date(),
    text: `Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu. Vivamus elementum nisl eu elit cursus, id luctus justo posuere. Vivamus elementum nisl eu elit cursus, id luctus justo posuere. Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. `,
    rating: 8,
    id: 5,
  },
  {
    user: {
      id: 8,
      name: `Denis`,
    },
    date: new Date(),
    text: `Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. Quisque fringilla lacinia pulvinar. Donec vel semper nunc, ac posuere lorem. Nulla sed tortor nunc.. `,
    rating: 3,
    id: 2,
  },
];

it(`TabReviews is rendered correctly`, () => {
  const tree = renderer.create((
    <TabReviews
      film={testFilms[0]}
      loadComments={() => {}}
      resetComments={() => {}}
      comments={comments}
    />
  ), {
    createNodeMock: () => {
      return {};
    }
  }).toJSON();

  expect(tree).toMatchSnapshot();
});
