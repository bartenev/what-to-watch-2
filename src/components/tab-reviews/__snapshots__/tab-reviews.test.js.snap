// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TabReviews is rendered correctly 1`] = `
<div
  className="movie-card__reviews movie-card__row"
>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu. Vivamus elementum nisl eu elit cursus, id luctus justo posuere. Vivamus elementum nisl eu elit cursus, id luctus justo posuere. Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. 
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Ivan
          </cite>
          <time
            className="review__date"
            dateTime="2021-11-01"
          >
            November 1, 2021
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        8
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. Quisque fringilla lacinia pulvinar. Donec vel semper nunc, ac posuere lorem. Nulla sed tortor nunc.. 
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Genka
          </cite>
          <time
            className="review__date"
            dateTime="2021-11-01"
          >
            November 1, 2021
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        0
      </div>
    </div>
  </div>
  <div
    className="movie-card__reviews-col"
  >
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Mauris interdum dolor a convallis pharetra. Fusce porttitor dictum mi, id tincidunt ligula lobortis eu. Integer commodo enim eget ullamcorper faucibus. Lorem ipsum dolor sit amet, consectetur adipiscing elit. 
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Cheba
          </cite>
          <time
            className="review__date"
            dateTime="2021-11-01"
          >
            November 1, 2021
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        5
      </div>
    </div>
    <div
      className="review"
    >
      <blockquote
        className="review__quote"
      >
        <p
          className="review__text"
        >
          Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. Praesent in nisi condimentum, aliquet nisl at, fermentum neque. Nunc sapien turpis, eleifend in odio at, ullamcorper efficitur lorem. 
        </p>
        <footer
          className="review__details"
        >
          <cite
            className="review__author"
          >
            Cheba
          </cite>
          <time
            className="review__date"
            dateTime="2021-11-01"
          >
            November 1, 2021
          </time>
        </footer>
      </blockquote>
      <div
        className="review__rating"
      >
        5
      </div>
    </div>
  </div>
</div>
`;
